boards/arm/nrf52/nrf52-feather/src/nrf52_ili9341_lcd.c

```diff
@@ -0,0 +1,410 @@
+ *
+ ****************************************************************************/
+
+static void nrf52_ili93414ws_select(FAR struct ili9341_lcd_s *lcd)
+{
+  FAR struct ili93414ws_lcd_s *priv = (FAR struct ili93414ws_lcd_s *)lcd;
+
+  SPI_LOCK(priv->spi, true);
+  SPI_SELECT(priv->spi, SPIDEV_DISPLAY(0), true);
+}
+
+/****************************************************************************
+ * Name: nrf52_ili93414ws_deselect
+ *
+ * Description:
+ *   De-select the SPI
+ *
+ * Input Parameters:
+ *   spi  - Reference to the public driver structure
+ *
+ * Returned Value:
+ *
+ ****************************************************************************/
+
+static void nrf52_ili93414ws_deselect(FAR struct ili9341_lcd_s *lcd)
+{
+  FAR struct ili93414ws_lcd_s *priv = (FAR struct ili93414ws_lcd_s *)lcd;
+
+  SPI_LOCK(priv->spi, false);
+  SPI_SELECT(priv->spi, SPIDEV_DISPLAY(0), false);
+}
+
+/****************************************************************************
+ * Name: nrf52_ili93414ws_sndcmd
+ *
+ * Description:
+ *   Send a command to the lcd driver.
+ *
+ * Input Parameters:
+ *   lcd  - Reference to the ili9341_lcd_s driver structure
+ *   cmd  - command to send
+ *
+ * Returned Value:
+ *   On success - OK
+ *
+ ****************************************************************************/
+
+static int nrf52_ili93414ws_sendcmd(
+    FAR struct ili9341_lcd_s *lcd, const uint8_t cmd)
+{
+  FAR struct ili93414ws_lcd_s *priv = (FAR struct ili93414ws_lcd_s *)lcd;
+
+  lcdinfo("%02x\n", cmd);
+
+  nrf52_gpio_write(ILI9341_DISPLAY_DC, false); /* Indicate CMD */
+  SPI_SEND(priv->spi, cmd);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: nrf52_ili93414ws_sendparam
+ *
+ * Description:
+ *   Send a parameter to the lcd driver.
+ *
+ * Input Parameters:
+ *   lcd    - Reference to the ili9341_lcd_s driver structure
+ *   param  - parameter to send
+ *
+ * Returned Value:
+ *   OK - On Success
+ *
+ ****************************************************************************/
+
+static int nrf52_ili93414ws_sendparam(FAR struct ili9341_lcd_s *lcd,
+                                      const uint8_t param)
+{
+  FAR struct ili93414ws_lcd_s *priv = (FAR struct ili93414ws_lcd_s *)lcd;
+
+  nrf52_gpio_write(ILI9341_DISPLAY_DC, true);  /* Indicate DATA */
+  SPI_SEND(priv->spi, param);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: nrf52_ili93414ws_sendgram
+ *
+ * Description:
+ *   Send a number of pixel words to the lcd driver gram.
+ *
+ * Input Parameters:
+ *   lcd    - Reference to the ili9341_lcd_s driver structure
+ *   wd     - Reference to the words to send
+ *   nwords - number of words to send
+ *
+ * Returned Value:
+ *   OK - On Success
+ *
+ ****************************************************************************/
+
+static int nrf52_ili93414ws_sendgram(FAR struct ili9341_lcd_s *lcd,
```

Comment:
@patacongo I for the gram send recv interfaces I had to add a byte swapping buffer because nrf52 SPI hardware does not support 16bit SPI.  If there was a way to tell the whole graphics stack to use BE instead of LE this could be avoided.  The buffer can be configured in size, but right now it is just set to the minimal of a single gram.  This also means that there is no real DMA advantage since the transfers are only 2 bytes at a time.

Reply:
> 
> 
> @patacongo for the gram send recv interfaces I had to add a byte swapping buffer because nrf52 SPI hardware does not support 16bit SPI. If there was a way to tell the whole graphics stack to use BE instead of LE for color this could be avoided. The buffer can be configured in size, but right now it is just set to the minimal of a single gram this also means that there is no real DMA advantage since the transfers are limited to the size of this buffer.

The location where the blitting is done is in graphics/nxglib/lcd.  But that uses macros defined in graphics/nxglib/nxglib_bitblit.h.  There you can see macros for:

    116 #if NXGLIB_BITSPERPIXEL < 8
    144 #elif NXGLIB_BITSPERPIXEL == 24
    188 #else /* NXGLIB_BITSPERPIXEL == 16 || NXGLIB_BITSPERPIXEL == 32 */
    221 #endif /* NXGLIB_BITSPERPIXEL */

You can see that for the 8, 16, and 32 bit version its just uses the native machine byte order (notice that the comment is wrong at line 188).  I think you would need to add and new configuration setting specifying the endianness of the RGB data.

I wouldn't just swap bytes, I would specifically ask for endian-ness because if this were used a big-endian machine that you would not want to swap.  The setting CONFIG_BIG_ENDIAN provides the native endian-ness.  You would swap only if !defined(CONFIG_BIG_ENDIAN) && defined(big-endian-LCD).

Reply:
> The location where the blitting is done is in graphics/nxglib/lcd. But that uses macros defined in graphics/nxglib/nxglib_bitblit.h.

Sorry, that is bad advice.  Those macros are only used by the framebuffer bit-blitters (graphics/nxglib/fb).  The blitting is hardcoded for the LCD in the files unser graphics/nxglib/lcd.

So you can't change the blitting in one place but rather in all of the 6 files at that location.  Sorry.

Reply:
Yeah it also will cause issues with other stacks like LVGL as well so it might be best to just leave the byteswap where it is for now. Most hardware supports 16bit mode, this just has a somewhat limited SPI controller.